### PR TITLE
Update CRUD API errors to match the new hierarchy

### PR DESCRIPTION
**Pull request by patrickfreed**
[SWIFT-144](https://jira.mongodb.org/browse/SWIFT-144) / [SWIFT-300](https://jira.mongodb.org/browse/SWIFT-300)

This PR updates reading CRUD operations to throw errors from the new error hierarchy. (SWIFT-144). As part of that, `InsertMany` now both throws a `.bulkWriteError` and is implemented as a bulk write (SWIFT-300). Also, command failure events have been updated to use the new errors as well.

### CODE REVIEW

File: Sources/MongoSwift/APM.swift [hunk 1]
```diff
@@ -132,7 +132,7 @@ public struct CommandFailedEvent: MongoEvent, InitializableFromOpaquePointer {
         self.commandName = String(cString: mongoc_apm_command_failed_get_command_name(event))
         var error = bson_error_t()
         mongoc_apm_command_failed_get_error(event, &error)
-        self.failure = MongoError.commandError(message: toErrorString(error))
+        self.failure = parseMongocError(error: error) // should always return a ServerError.commandError
```
Comment (mbroadst) [inline Sources/MongoSwift/APM.swift]: nit: should `parseMongocError` s first parameter not be named? Seems redundant to have `error: error` everywhere, since the method also tells us its parsing an error

File: Sources/MongoSwift/MongoClient.swift [hunk 2]
```diff
@@ -150,6 +150,7 @@ public class MongoClient {
         if options?.eventMonitoring == true { self.initializeMonitoring() }
 
         guard mongoc_client_set_error_api(self._client, MONGOC_ERROR_API_VERSION_2) else {
+            close()
```
Comment (mbroadst) [inline Sources/MongoSwift/MongoClient.swift]: nice catch

File: Sources/MongoSwift/BSON/BSONDecoder.swift [hunk 3]
```diff
@@ -164,6 +164,20 @@ public class BSONDecoder {
                                   debugDescription: "Unable to parse JSON string \(json)"))
     }
 
+    /// Internal function used to  convert encountered `DecodingError`s to `.internalErrors`. Use this when using the
```
Comment (kmahar) [inline Sources/MongoSwift/BSON/BSONDecoder.swift]: the first sentence reads like this is something that takes in a `DecodingError` and returns an `internalError`. can we swap the sentences or rephrase so it's clearer that it is basically just `decode` with a different error type?

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift [hunk 4]
```diff
@@ -53,12 +56,17 @@ extension MongoCollection {
         }
 
         /// Adds the `deleteOne` operation to a bulk write
+        ///
+        /// - Throws:
+        ///   - `UserError.invalidArgumentError` if the options form an invalid combination.
+        ///   - `EncodingError` if an error occurs while encoding the options to BSON.
         public func addToBulkWrite(bulk: BulkWriteOperation, index: Int) throws {
-            let opts = try BSONEncoder().encode(self.options)
+            let opts = try BSONEncoder().encode(options)
```
Comment (mbroadst) [inline Sources/MongoSwift/MongoCollection+BulkWrite.swift]: I thought we _were_ doing explicit `self`?

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift [hunk 5]
```diff
@@ -53,12 +56,17 @@ extension MongoCollection {
         }
 
         /// Adds the `deleteOne` operation to a bulk write
+        ///
```
Comment (kmahar) [inline Sources/MongoSwift/MongoCollection+BulkWrite.swift]: nit: I think @jmikola and I decided a while back that for > 3 lines docstrings we should use `/** ... */` instead

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift [hunk 6]
```diff
@@ -81,12 +89,16 @@ extension MongoCollection {
...
+        /// - Throws:
+        ///   - `UserError.invalidArgumentError` if the options form an invalid combination.
+        ///   - `EncodingError` if an error occurs while encoding the options to BSON.
         public func addToBulkWrite(bulk: BulkWriteOperation, index: Int) throws {
             let opts = try BSONEncoder().encode(options)
             var error = bson_error_t()
 
             guard mongoc_bulk_operation_remove_many_with_opts(bulk.bulk, self.filter.data, opts.data, &error) else {
-                throw MongoError.invalidArgument(message: toErrorString(error))
+                throw UserError.invalidArgumentError(message: toErrorString(error))
```
Comment (mbroadst) [inline Sources/MongoSwift/MongoCollection+BulkWrite.swift]: Why isn't this line using `parseMongocError`?  It seems like there are a few more cases like this, I'll restrict my comment to this one

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift [hunk 7]
```diff
@@ -239,6 +263,10 @@ extension MongoCollection {
         }
 
         /// Adds the `updateMany` operation to a bulk write
+        ///
```
Comment (kmahar) [inline Sources/MongoSwift/MongoCollection+BulkWrite.swift]: same nit as above about `///` vs `/** ... */`

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift [hunk 8]
```diff
@@ -12,20 +12,23 @@ extension MongoCollection {
...
+            throw UserError.invalidArgumentError(message: "requests cannot be empty")
         }
 
         let opts = try BSONEncoder().encode(options)
-        let bulk = BulkWriteOperation(collection: self._collection, opts: opts?.data)
+        let bulk = BulkWriteOperation(collection: self._collection, opts: opts)
 
         try requests.enumerated().forEach { index, model in
             try model.addToBulkWrite(bulk: bulk, index: index)
+            bulk.requests[index] = model
```
Comment (mbroadst) [inline Sources/MongoSwift/MongoCollection+BulkWrite.swift]: Can you explain why this change was needed?

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift (hunk 4 above)
Comment (kmahar) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: the swiftlint rule seems to miss a lot of these... 😩

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift [hunk 9]
```diff
@@ -197,7 +217,11 @@ extension MongoCollection {
             self.options = UpdateModelOptions(arrayFilters: arrayFilters, collation: collation, upsert: upsert)
         }
 
-        /// Adds the `updateOne` operation to a bulk write
+        /// Adds the `updateOne` operation to a bulk write.
```
Comment (kmahar) [inline Sources/MongoSwift/MongoCollection+BulkWrite.swift]: same nit as above about /// vs /** ... */

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift [hunk 10]
```diff
@@ -393,21 +433,30 @@ public struct BulkWriteResult {
...
+    /// Internal initializer used for testing purposes and error handling.
+    internal init(
+            deletedCount: Int? = nil,
+            insertedCount: Int? = nil,
+            insertedIds: [Int: BSONValue]? = nil,
+            matchedCount: Int? = nil,
+            modifiedCount: Int? = nil,
+            upsertedCount: Int? = nil,
+            upsertedIds: [Int: BSONValue]? = nil) {
+        self.deletedCount = deletedCount ?? 0
```
Comment (kmahar) [inline Sources/MongoSwift/MongoCollection+BulkWrite.swift]: would it make sense to just use all of these default values as the defaults for the parameters, rather than setting new default values inside here?

File: Sources/MongoSwift/BSON/BSONDecoder.swift (hunk 3 above)
Comment (patrickfreed) [reply Sources/MongoSwift/BSON/BSONDecoder.swift]: Fixed. Now reads:
>Internal version of `decode` that throws `.internalErrors` instead of `DecodingErrors`. Use this when using the decoder internally on non-user-modified types.

File: Sources/MongoSwift/BSON/BSONDecoder.swift [hunk 11]
```diff
@@ -164,6 +164,20 @@ public class BSONDecoder {
...
+    /// decoder internally on non-user-modified types.
+    public func internalDecode<T: Decodable>(
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/BSON/BSONDecoder.swift]: Yep, fixed.

File: Sources/MongoSwift/MongoClient.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/MongoClient.swift]: fixed

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift (hunk 5 above)
Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: Fixed.

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift (hunk 9 above)
Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: fixed

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift (hunk 7 above)
Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: fixed

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift (hunk 10 above)
Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: I would still have to unwrap them even if they had default values. I could force unwrap and add a swiftlint ignore if you think that's preferable.

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift (hunk 8 above)
Comment (mbroadst) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: Where does the requirement to access the requests in the error come from?

Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: The definition of `BulkWriteError` in the [crud spec](https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#error-handling) includes it.

Comment (mbroadst) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: Well that would certainly justify it! I still think we should try to offload this onto the C driver, we shouldn't have to allocate these twice for potential error scenarios. As it stands, I'm paying the price for something I'll hopefully never see.

File: Sources/MongoSwift/MongoError.swift [hunk 12]
```diff
@@ -169,6 +172,95 @@ internal func parseMongocError(error: bson_error_t, errorLabels: [String]? = nil
     }
 }
 
+/// Internal function used to get an appropriate error from a server reply to a command.
+internal func getErrorFromReply(
```
Comment (kmahar) [inline Sources/MongoSwift/MongoError.swift]: this might be a bit more readable if we broke out the bulk write and write error portions into separate methods.

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift (hunk 8 above)
Comment (jmikola) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: IIRC, including the write model in the error is optional. The only requirement is that we report the index for individual write errors. I spoke to @patrickfreed about this last week as I saw he added this behavior but I didn't realize at the time that it would add quite a bit of overhead to the common no-error case.

I don't see how this can be offloaded to the C driver, as I don't believe it provides access to operations once they're added to a `mongoc_bulk_operation_t`. Might be best just to remove this altogether.

Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: The write model field in the error is an optional type (meaning it wouldn't be present in some cases, I guess. I'm not sure what those cases would be, though.), but it is _not_ marked as "NOT REQUIRED" as the fields which are optional for drivers to implement are. It seems, according to spec at least, the field is required. Whether or not it is worth it to us to conform to the spec in this case is another question, though.

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift [hunk 13]
```diff
@@ -52,13 +55,20 @@ extension MongoCollection {
             self.options = DeleteModelOptions(collation: collation)
         }
 
-        /// Adds the `deleteOne` operation to a bulk write
+        /**
+         * Adds the `deleteOne` operation to a bulk write
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: Done

File: Sources/MongoSwift/MongoError.swift (hunk 12 above)
Comment (patrickfreed) [reply Sources/MongoSwift/MongoError.swift]: done

File: Sources/MongoSwift/MongoClient.swift [hunk 14]
```diff
@@ -150,6 +150,7 @@ public class MongoClient {
         if options?.eventMonitoring == true { self.initializeMonitoring() }
 
         guard mongoc_client_set_error_api(self._client, MONGOC_ERROR_API_VERSION_2) else {
+            self.close()
```
Comment (jmikola) [inline Sources/MongoSwift/MongoClient.swift]: Is this done in other cases, such as failure to initialize monitoring?

Was this added due to code analysis, or would we actually expect `mongoc_client_set_error_api()` to fail in a real use case?

File: Sources/MongoSwift/MongoCollection+Indexes.swift [hunk 15]
```diff
@@ -233,9 +251,11 @@ extension MongoCollection {
 
         let opts = try encoder.encode(options)
         var error = bson_error_t()
+        let reply = Document()
```
[the remark this answers is not in the document]
Comment (jmikola) [reply Sources/MongoSwift/MongoCollection+Indexes.swift]: I don't think that was a problem. On error, you were still reporting the error string from libmongoc's `bson_error_t`. With this change, you're now surfacing the original command error (if any).

The reply was never really needed, since the success case is just returning a list of index names, which can be derived from the input models.

File: Sources/MongoSwift/MongoClient.swift (hunk 14 above)
Comment (patrickfreed) [reply Sources/MongoSwift/MongoClient.swift]: As the driver is now, this call should never fail. If we change our threading/client pooling model, it will (you need to invoke this on the pool rather than a client derived from the pool). I added the check so that if we did start using client pools, tests would break unless this call was updated.

File: Sources/MongoSwift/MongoCollection+BulkWrite.swift (hunk 8 above)
Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: Python has a `dict` that represents a `BulkWriteErrror` which has an `'op'` field in it. It looks like Node has similar behavior.

Java, C#, Go, and Ruby don't report the request.

So none of the drivers report an actual `WriteModel`, but Node and Python do report some information about the request that caused the error.

Comment (mbroadst) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: My thought about offloading onto the C driver was that we ask for an accessor for `mongoc_bulk_operation_t` operations. Outside of that, I'd argue that we amend the spec to clarify that this is optional, and not implement it until someone asks for it.

Comment (kmahar) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: I agree with @mbroadst

Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: To summarize:
I'll remove the request storage from `BulkWriteError` and file a follow-up SPEC ticket regarding making it NOT REQUIRED. If in the future libmongoc provides some sort of accessor on `bulk_operation_t` for getting operations by index or otherwise reports them in the errors, we can revisit this.

Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+BulkWrite.swift]: Filed [SPEC-1209](https://jira.mongodb.org/browse/SPEC-1209)
